### PR TITLE
Added server/process crash resiliency

### PR DESCRIPTION
**Pull request by RyanVasichko**
LiteJob is currently lacking crash resiliency. Its mechanism for popping items from the queue deletes the item from the queue table, resulting in a potential lost job if the server/process crashes before the job is completed. 

This PR addresses that issue by:

1. Adding a workers table
2. Associating each queue entry with a worker whenever it is popped and modifying the `pop` logic to ignore queues that have been assigned to a worker
3. Adding a heartbeat for each worker
4. Periodically clearing workers from the workers table who have missed a configurable number of heartbeats
5. Periodically setting the `worker_id` to `NULL` for queues that are referencing workers that no longer exist in the worker table, which will result in reprocessing

### CODE REVIEW

**Comment by RyanVasichko** [comment]
@oldmoe  - not sure if you get notified for PRs or not, but pinging you just in case

**Comment by oldmoe** [comment]
Hi Ryan, thanks for the PR, some thoughts below:

I was originally planning to make pops an update rather than a delete, but I was not thinking of capturing worker ids and heart beats, instead I am envisioning a simpler scheme of time out for jobs in the running queue, after which they can be either pushed back to the main queue to be retried or put in the dead queue if they exhausted their chances. There are multiple reasons for that.

1. Minimize per worker overhead, since there could many workers processing jobs, in multiple processes
2. This will also be spawn mode compatible, since this modes creates "anonymous" sub workers
3. Allowing the fetch of multiple jobs at once (an optimization I am planning in a later release) when multiple workers are idle
4. The timeout is more like a property of the job/queue rather than of the worker
5. Timed out jobs can still be easily garbage collected without much overhead on the system, specially no increased writes

In summary, I see the need for crash resilience and I am hoping to get that in, but I would rather keep it simple as much as possible. Would love to have your thoughts as well, and I am happy to work with you on the changes if you think my proposal makes sense.

**Comment by RyanVasichko** [comment]
I had originally thought of checking for long-running jobs and counting jobs that had been running for too long as dead, but I wasn't in love with the idea of possibly re-running a long running job, even though ideally a job shouldn't take long enough that it might be considered dead.

I think the performance impact should be fairly minimal since the worker heartbeat is running in a separate thread and is only waking up every 60 seconds to do a quick checkin. Although I think bumping the default heartbeat interval up to 5  or even 10 minutes makes sense. And the abandoned job rescuer is running every 10 minutes or so and just does a quick `JOIN` and `UPDATE` so I doubt that would really have much of an impact on performance. And with a timeout system we'd still have to check for timed out jobs and do an `UPDATE` so it's really the same either way.

This solution should also be spawn-mode compatible as far as I can tell, since in spawn mode the worker isn't really spawning a separate worker, it's just forking the job processing into a separate fiber/thread while the worker stays alive. 

I think a heartbeat system is more robust and the performance impact should be negligible, with the only downside being the inability to pop multiple jobs at once. Although I'm not sure that we really gain all that much from being able to do that. That said I'm not married to this solution, so if you'd prefer to just consider a job "dead" if a certain amount of time has passed and it's still in the processing queue I'm game for going with that approach.

**Comment by gshaw** [comment]
> if you'd prefer to just consider a job "dead" if a certain amount of time has passed and it's still in the processing queue I'm game for going with that approach.

If this approach is used it would be very useful for jobs to be able to indicate their maximum run time.  E.g., most jobs will execute very quickly but I've had jobs on other systems take 10s of minutes to perform.

**Comment by RyanVasichko** [comment]
Closing as this is replaced by #67 

@gshaw - that's a good suggestion and I'll implement it after #67 is completed.
